pkg/fuzzer/fuzzer.go

```diff
@@ -0,0 +1,307 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package fuzzer
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/syzkaller/pkg/hash"
+	"github.com/google/syzkaller/pkg/ipc"
+	"github.com/google/syzkaller/pkg/rpctype"
+	"github.com/google/syzkaller/prog"
+)
+
+type Fuzzer struct {
+	Config    *Config
+	Corpus    *Corpus
+	NewInputs chan rpctype.Input
+
+	ctx    context.Context
+	mu     sync.Mutex
+	stats  map[string]uint64
+	rnd    *rand.Rand
+	target *prog.Target
+
+	choiceTable *choiceTableProxy
+
+	nextExec     *priorityQueue[*Request]
+	runningExecs map[*Request]time.Time
+	jobIDs       map[job]int64
+	nextJobID    int64
+
+	queuedCandidates atomic.Uint64
+	// If the source of candidates runs out of them, we risk
+	// generating too many needCandidate requests (one for
+	// each Config.MinCandidates). We prevent this with candidatesRequested.
+	candidatesRequested atomic.Bool
+}
+
+func NewFuzzer(ctx context.Context, cfg *Config, rnd *rand.Rand,
+	target *prog.Target) *Fuzzer {
+	f := &Fuzzer{
+		Config: cfg,
+		Corpus: newCorpus(),
+
+		ctx:          ctx,
+		stats:        map[string]uint64{},
+		rnd:          rnd,
+		target:       target,
+		choiceTable:  &choiceTableProxy{},
+		nextExec:     makePriorityQueue[*Request](),
+		runningExecs: map[*Request]time.Time{},
+		jobIDs:       map[job]int64{},
+	}
+	go func() {
+		<-ctx.Done()
+		close(f.NewInputs)
```

Comment:
This is racy and can panic. It's not OK to close a channel concurrently with send operations.
If there is only 1 goroutine that sends, then it needs to close the channel. If there are multiple, then it's OK close only after waiting for all sending goroutines.

Reply:
Ah, sorry, I forgot to push a change that deletes this goroutine. Now `NewInputs` is just part of the `Config`.